### ISSUE TITLE
Guidelines for CPU and memory optimizations

### ISSUE THREAD

**Issue by fmunteanu** [thumbor/thumbor#1053]
Hi,

First, let me start by thanking you for this beautiful and elegant solution. We are using Thumbor for last 6 months in production, serving on average 10,000,000 images/hour with zero issues. Now that we know the product is very stable, we would like to optimize the EC2 instances to further enhance Thumbor on processing more images/minute.

We are using a Thumbor stack deployed with AWS CloudFormation and an autoscaling group of c5.2xlarge CentOS7 EC2 instances, fully optimized to take advantage of the 10Gbit network infrastructure AWS offers. All images are cached with Akamai and whatever missed, is cached locally for an hour into a Redis cache.

Current infrastructure (image served with Thumbor🥇):
![Thumbor Infrastructure](https://thumbnails.trvl-media.com/pMwvGSoaNIJ5onfXkYc8crR-2lg=/assets/nucleus/thumbor.jpg)
The thumbor-6.4.2 RPM is available into AXIVO repository. We use [python-pillow-simd-4.3.0](https://github.com/uploadcare/pillow-simd), as is 2.5 times faster compared to python-pillow-4.3.0 and proxy all calls through nginx-1.13.10. This allowed us to reduce the CPU usage by approximately 20%.

While CPU is currently floating around 75% values on each instance, I noticed the memory is partially used. This is the Thumbor configuration we use, on 8 processors and 16GB of RAM:
```
$ cat /etc/thumbor/thumbor.conf
ALLOW_OLD_URLS = False
ALLOW_UNSAFE_URL = False
AUTO_WEBP = True
DETECTORS = [
	'thumbor.detectors.face_detector',
	'thumbor.detectors.feature_detector',
	'thumbor.detectors.glasses_detector',
	'thumbor.detectors.profile_detector'
]
OPTIMIZERS = [
	'thumbor.optimizers.jpegtran'
]
ENGINE_THREADPOOL_SIZE = 16
JPEGTRAN_PATH = '/bin/jpegtran'
HTTP_LOADER_MAX_CLIENTS = 16384
MAX_AGE = 2592000
MAX_AGE_TEMP_IMAGE = 60
STORAGE = 'tc_redis.storages.redis_storage'
STORAGE_EXPIRATION_SECONDS = 3600
REDIS_STORAGE_IGNORE_ERRORS = True
REDIS_STORAGE_SERVER_HOST = '0001.usw2.cache.amazonaws.com'
REDIS_STORAGE_SERVER_PORT = 6379
REDIS_STORAGE_SERVER_DB = 0
REDIS_STORAGE_SERVER_PASSWORD = None
RESULT_STORAGE = 'tc_redis.result_storages.redis_result_storage'
RESULT_STORAGE_EXPIRATION_SECONDS = 3600
REDIS_RESULT_STORAGE_IGNORE_ERRORS = True
REDIS_RESULT_STORAGE_SERVER_HOST = '0001.usw2.cache.amazonaws.com'
REDIS_RESULT_STORAGE_SERVER_PORT = 6379
REDIS_RESULT_STORAGE_SERVER_DB = 1
REDIS_RESULT_STORAGE_SERVER_PASSWORD = None

$ free -m
              total        used        free      shared  buff/cache   available
Mem:          15309        7604        5871         284        1834        7024
Swap:             0           0           0

$ systemctl --type=service --state=running list-units | grep thumbor
thumbor@8000.service             loaded active running Thumbor (thumbor@8000)
thumbor@8001.service             loaded active running Thumbor (thumbor@8001)
thumbor@8002.service             loaded active running Thumbor (thumbor@8002)
thumbor@8003.service             loaded active running Thumbor (thumbor@8003)
thumbor@8004.service             loaded active running Thumbor (thumbor@8004)
thumbor@8005.service             loaded active running Thumbor (thumbor@8005)
thumbor@8006.service             loaded active running Thumbor (thumbor@8006)
thumbor@8007.service             loaded active running Thumbor (thumbor@8007)
```
The `ENGINE_THREADPOOL_SIZE` value is `2 x nproc`.

Are there any other settings I should take advantage of, to optimize the CPU usage and also increase the memory usage? @heynemann, @guilhermef, @cezarsa, your input is greatly appreciated on this matter.

Regards,

Floren Munteanu

**Comment by heynemann** [comment]
Hmmm... Are you using smart detection? With OpenCV?

If so, I'd seriously advise you to use the queued detector. We haver RemoteCV for that purpose (a queue that uses a Resque port to python). We NEVER do Face Detection synchronously.

When a new image arrives with `/smart` in it, we send it to the remote cv queue (unless it has been detected before, the detection result can also be stored in redis). Then we serve the image request with a very low expiration (meaning the caller will try again shortly).

The next time the image is requested it will likely have the detection results. We then serve it with the proper expiration.

This has significantly reduced the CPU pressure on our farm.

Hope it helps!

**Comment by heynemann** [comment]
Btw thank you VERY much for your kind words. We really appreciate it. It is good to hear that it has helped you as much as it has helped us and that baking extensibility into Thumbor, while not a primary concern for Globo.com has payed off many times by allowing us to collaborate with so many talented people! Thank you!

**Comment by fmunteanu** [comment]
@heynemann, we use the /smart switch in all URLs. Can you detail further your RemoteCV setup to alleviate the CPU pressure? I looked few months ago at RemoteCV and had very specific software dependencies with version already deprecated into CentOS7. I'll have to revisit this, since is really important we reduce the CPU load.

These are current dependencies on CentOS7 thumbor-6.4.2 RPM:
```
Requires:		opencv-python
Requires:		python2-argparse >= 1.2.1
Requires:		python-derpconf >= 0.2.0
Requires:		python-futures
Requires:		python-libthumbor >= 1.3.2
Requires:		python-piexif >= 1.0.13
Requires:		python-piexif < 1.1.0
Requires:		python-pillow-simd >= 3.0.0
Requires:		python-pillow-simd < 5.0.0
Requires:		python-pycryptodomex >= 3.4.7
Requires:		python-pycurl >= 7.19.0
Requires:		python-pycurl < 7.44.0
Requires:		python-schedule
Requires:		python-six
Requires:		python-statsd >= 3.0.1
Requires:		python-tornado >= 4.1.0
Requires:		python-tornado < 5.0.0
Requires:		python-webcolors
Requires:		pytz
Requires:		rsyslog
```

**Comment by heynemann** [comment]
We might work to upgrade RemoteCV to newer dependencies if it is something that would help you.

Well, we don't run any detection algorithms in the thumbor processes. We delegate those to other machines whose sole purpose is detecting faces and objects. This way the thumbor machines (lots of them) can process images and only images, while the remotecv machines (few) can do face/feature detection.

This way your users get the best of both worlds: performance and quality. They get faster image renderings since you don't have to do face detection while rendering the image. On the other hand they get the best cropped image with face detection data eventually, thus improving quality gradually. It's an eventually consistent way of working with smart detection.

**Comment by heynemann** [comment]
Maybe this helps: http://heynemann.logdown.com/posts/152158-scaling-thumbor

**Comment by fmunteanu** [comment]
@heynemann, there is only one dependency version to allow, python2-argparse-1.4.0 instead of 1.3.0, we use it already in Thumbor RPM. python2-pyres is fine, not a conflict.

We do very little detection of any kind, do DETECTORS slowdown everything? If yes, I had no idea and we will spin a separate service for that, with RemoteCV. I'm going to look into your blog details.

As you can see from above thumbor.conf, we use 2 Redis databases for images storage. Should I cache the actual images locally? So far, Redis helps a lot to serve images from cache.

**Comment by fmunteanu** [comment]
Proposed configuration for cluster using /smart switch:
```
$ cat /etc/thumbor/thumbor.conf
ALLOW_OLD_URLS = False
ALLOW_UNSAFE_URL = False
AUTO_WEBP = True
OPTIMIZERS = [
	'thumbor.optimizers.jpegtran'
]
ENGINE_THREADPOOL_SIZE = 16
JPEGTRAN_PATH = '/bin/jpegtran'
HTTP_LOADER_MAX_CLIENTS = 16384
MAX_AGE = 2592000
MAX_AGE_TEMP_IMAGE = 60
STORAGE = 'tc_redis.storages.redis_storage'
STORAGE_EXPIRATION_SECONDS = 3600
REDIS_STORAGE_IGNORE_ERRORS = True
REDIS_STORAGE_SERVER_HOST = '0001.usw2.cache.amazonaws.com'
REDIS_STORAGE_SERVER_PORT = 6379
REDIS_STORAGE_SERVER_DB = 0
REDIS_STORAGE_SERVER_PASSWORD = None
RESULT_STORAGE = 'tc_redis.result_storages.redis_result_storage'
RESULT_STORAGE_EXPIRATION_SECONDS = 3600
REDIS_RESULT_STORAGE_IGNORE_ERRORS = True
REDIS_RESULT_STORAGE_SERVER_HOST = '0001.usw2.cache.amazonaws.com'
REDIS_RESULT_STORAGE_SERVER_PORT = 6379
REDIS_RESULT_STORAGE_SERVER_DB = 1
REDIS_RESULT_STORAGE_SERVER_PASSWORD = None
```
Let me know if there is something I missed.

**Comment by heynemann** [comment]
I think you now configured not to use any detectors, which is not good IMHO. You'll be using thumbor's default value for it which is an empy list. You essentially turned off smart cropping.

In the post above I detailed using remotecv from thumbor's perspective. Is there anything you think I could improve upon? Any specific doubts?

Detectors do not slow down everything, but detecting faces in a large (4k) image can be very slow (many seconds). Even if it does not take long, computer vision is a very cpu intensive operation. That's why we have specific servers for them. Servers that do not need to deal with image compression, http(s) and all the request logic that thumbor already deals with. All they need to do is get an image from somewhere and detect objects in that image, while storing the results somewhere.

Does that make sense?

**Comment by kkopachev** [comment]
@fmunteanu Is using redis for storage any faster then retrieving image from s3 through loader/result_storage? How big is redis instances?

**Comment by fmunteanu** [comment]
@heynemann it does, thanks for clarifications. I started working on separate cluster powered by RemoteCV and OpenCV3 but there are several deprecated functions in your software. I was wondering if you could merge [pull 32](https://github.com/thumbor/opencv-engine/pull/32)? That will fix all my issues.

@kkopachev we tried S3 storage but the API calls are too expensive (10 million calls/hour). Elasticache Redis is very fast and it does not add additional costs, you pay only for the memory usage. Still, we are moving away from it, to a Redis cluster running on EC2 instances and managed with Sentinel. We are still waiting for AWS Fargate to be deployed in all regions.

**Comment by fmunteanu** [comment]
The opencv-3.4.1 rpms are now available into AXIVO repository.
```
Available Packages
opencv.x86_64                          3.4.1-1.el7                axivo
opencv-contrib.x86_64                  3.4.1-1.el7                axivo
opencv-core.x86_64                     3.4.1-1.el7                axivo
opencv-devel.x86_64                    3.4.1-1.el7                axivo
python2-opencv.x86_64                        3.4.1-1.el7         axivo
python2-thumbor-opencv-engine.noarch         1.0.1-1.el7         axivo
```
Awaiting on the merge to fix `python2-thumbor-opencv-engine` rpm.

**Comment by fmunteanu** [comment]
Just to clarify, `python2-thumbor-opencv-engine` rpm is currently functional. I extracted only the cv3 engine from @Arthien 's PR#32 and applied it as patch to existing opencv-engine-1.0.1 release. I was referring to the code merge into a new version. @heynemann, see below the patch:
```
 opencv_engine/__init__.py   |   2 +-
 opencv_engine/engine_cv3.py | 234 ++++++++++++++++++++++++++++++++++++++++++++
 setup.py                    |   2 +
 3 files changed, 237 insertions(+), 1 deletion(-)
 create mode 100644 opencv_engine/engine_cv3.py

diff --git a/opencv_engine/__init__.py b/opencv_engine/__init__.py
index 1647d10..dfb5a57 100644
--- a/opencv_engine/__init__.py
+++ b/opencv_engine/__init__.py
@@ -6,6 +6,6 @@ import logging
 __version__ = '1.0.1'

 try:
-    from opencv_engine.engine import Engine  # NOQA
+    from opencv_engine.engine_cv3 import Engine  # NOQA
 except ImportError:
     logging.exception('Could not import opencv_engine. Probably due to setup.py installing it.')
diff --git a/opencv_engine/engine_cv3.py b/opencv_engine/engine_cv3.py
new file mode 100644
index 0000000..1d22b9e
--- /dev/null
+++ b/opencv_engine/engine_cv3.py
@@ -0,0 +1,234 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Licensed under the MIT license:
+# http://www.opensource.org/licenses/mit-license
+# Copyright (c) 2016 fanhero.com christian@fanhero.com
+
+import cv2
+import numpy as np
+
+from colour import Color
+from thumbor.engines import BaseEngine
+from pexif import JpegFile, ExifSegment
+
+try:
+    from thumbor.ext.filters import _composite
+    FILTERS_AVAILABLE = True
+except ImportError:
+    FILTERS_AVAILABLE = False
+
+FORMATS = {
+    '.jpg': 'JPEG',
+    '.jpeg': 'JPEG',
+    '.gif': 'GIF',
+    '.png': 'PNG',
+    '.webp': 'WEBP'
+}
+
+
+class Engine(BaseEngine):
+    @property
+    def image_depth(self):
+        if self.image is None:
+            return np.uint8
+        return self.image.dtype
+
+    @property
+    def image_channels(self):
+        if self.image is None:
+            return 3
+        # if the image is grayscale
+        try:
+            return self.image.shape[2]
+        except IndexError:
+            return 1
+
+    @classmethod
+    def parse_hex_color(cls, color):
+        try:
+            color = Color(color).get_rgb()
+            return tuple(c * 255 for c in reversed(color))
+        except Exception:
+            return None
+
+    def gen_image(self, size, color_value):
+        if color_value == 'transparent':
+            color = (255, 255, 255, 255)
+            img = np.zeros((size[1], size[0], 4), self.image_depth)
+        else:
+            img = np.zeros((size[1], size[0], self.image_channels), self.image_depth)
+            color = self.parse_hex_color(color_value)
+            if not color:
+                raise ValueError('Color %s is not valid.' % color_value)
+        img[:] = color
+        return img
+
+    def create_image(self, buffer):
+        # FIXME: opencv doesn't support gifs, even worse, the library
+        # segfaults when trying to decoding a gif. An exception is a
+        # less drastic measure.
+        try:
+            if FORMATS[self.extension] == 'GIF':
+                raise ValueError("opencv doesn't support gifs")
+        except KeyError:
+            pass
+
+        img = cv2.imdecode(np.frombuffer(buffer, np.uint8), -1)
+        if FORMATS[self.extension] == 'JPEG':
+            self.exif = None
+            try:
+                info = JpegFile.fromString(buffer).get_exif()
+                if info:
+                    self.exif = info.data
+                    self.exif_marker = info.marker
+            except Exception:
+                pass
+        return img
+
+    @property
+    def size(self):
+        return self.image.shape[1], self.image.shape[0]
+
+    def normalize(self):
+        pass
+
+    def resize(self, width, height):
+        r = height / self.size[1]
+        width = int(self.size[0] * r)
+        dim = (int(round(width, 0)), int(round(height, 0)))
+        self.image = cv2.resize(self.image, dim, interpolation=cv2.INTER_AREA)
+
+    def crop(self, left, top, right, bottom):
+        self.image = self.image[top: bottom, left: right]
+
+    def rotate(self, degrees):
+        # see http://stackoverflow.com/a/23990392
+        if degrees == 90:
+            self.image = cv2.transpose(self.image)
+            cv2.flip(self.image, 0, self.image)
+        elif degrees == 180:
+            cv2.flip(self.image, -1, self.image)
+        elif degrees == 270:
+            self.image = cv2.transpose(self.image)
+            cv2.flip(self.image, 1, self.image)
+        else:
+            # see http://stackoverflow.com/a/37347070
+            # one pixel glitch seems to happen with 90/180/270
+            # degrees pictures in this algorithm if you check
+            # the typical github.com/recurser/exif-orientation-examples
+            # but the above transpose/flip algorithm is working fine
+            # for those cases already
+            width, height = self.size
+            image_center = (width / 2, height / 2)
+            rot_mat = cv2.getRotationMatrix2D(image_center, degrees, 1.0)
+
+            abs_cos = abs(rot_mat[0, 0])
+            abs_sin = abs(rot_mat[0, 1])
+            bound_w = int((height * abs_sin) + (width * abs_cos))
+            bound_h = int((height * abs_cos) + (width * abs_sin))
+
+            rot_mat[0, 2] += ((bound_w / 2) - image_center[0])
+            rot_mat[1, 2] += ((bound_h / 2) - image_center[1])
+
+            self.image = cv2.warpAffine(self.image, rot_mat, (bound_w, bound_h))
+
+    def flip_vertically(self):
+        self.image = np.flipud(self.image)
+
+    def flip_horizontally(self):
+        self.image = np.fliplr(self.image)
+
+    def read(self, extension=None, quality=None):
+        if quality is None:
+            quality = self.context.config.QUALITY
+
+        options = None
+        extension = extension or self.extension
+        try:
+            if FORMATS[extension] == 'JPEG':
+                options = [cv2.IMWRITE_JPEG_QUALITY, quality]
+        except KeyError:
+            # default is JPEG so
+            options = [cv2.IMWRITE_JPEG_QUALITY, quality]
+
+        try:
+            if FORMATS[extension] == 'WEBP':
+                options = [cv2.IMWRITE_WEBP_QUALITY, quality]
+        except KeyError:
+            options = [cv2.IMWRITE_JPEG_QUALITY, quality]
+
+        success, buf = cv2.imencode(extension, self.image, options or [])
+        data = buf.tostring()
+
+        if FORMATS[extension] == 'JPEG' and self.context.config.PRESERVE_EXIF_INFO:
+            if hasattr(self, 'exif') and self.exif != None:
+                img = JpegFile.fromString(data)
+                img._segments.insert(0, ExifSegment(self.exif_marker, None, self.exif, 'rw'))
+                data = img.writeString()
+
+        return data
+
+    def set_image_data(self, data):
+        self.image = np.frombuffer(data, dtype=self.image.dtype).reshape(self.image.shape)
+
+    def image_data_as_rgb(self, update_image=True):
+        if self.image_channels == 4:
+            mode = 'BGRA'
+        elif self.image_channels == 3:
+            mode = 'BGR'
+        else:
+            mode = 'BGR'
+            rgb_copy = np.zeros((self.size[1], self.size[0], 3), self.image.dtype)
+            cv2.cvtColor(self.image, cv2.COLOR_GRAY2BGR, rgb_copy)
+            self.image = rgb_copy
+        return mode, self.image.tostring()
+
+    def draw_rectangle(self, x, y, width, height):
+        cv2.rectangle(self.image, (int(x), int(y)), (int(x + width), int(y + height)), (255, 255, 255))
+
+    def convert_to_grayscale(self, update_image=True, with_alpha=True):
+        image = None
+        if self.image_channels >= 3 and with_alpha:
+            image = cv2.cvtColor(self.image, cv2.COLOR_BGRA2GRAY)
+        elif self.image_channels >= 3:
+            image = cv2.cvtColor(self.image, cv2.COLOR_BGR2GRAY)
+        elif self.image_channels == 1:
+            # Already grayscale,
+            image = self.image
+        if update_image:
+            self.image = image
+        elif self.image_depth == np.uint16:
+            #Feature detector reqiures uint8 images
+            image = np.array(image, dtype='uint8')
+        return image
+
+    def paste(self, other_engine, pos, merge=True):
+        if merge and not FILTERS_AVAILABLE:
+            raise RuntimeError(
+                'You need filters enabled to use paste with merge. Please reinstall ' +
+                'thumbor with proper compilation of its filters.')
+
+        self.enable_alpha()
+        other_engine.enable_alpha()
+
+        sz = self.size
+        other_size = other_engine.size
+
+        mode, data = self.image_data_as_rgb()
+        other_mode, other_data = other_engine.image_data_as_rgb()
+
+        imgdata = _composite.apply(
+            mode, data, sz[0], sz[1],
+            other_data, other_size[0], other_size[1], pos[0], pos[1], merge)
+
+        self.set_image_data(imgdata)
+
+    def enable_alpha(self):
+        if self.image_channels < 4:
+            with_alpha = np.zeros((self.size[1], self.size[0], 4), self.image.dtype)
+            if self.image_channels == 3:
+                cv2.cvtColor(self.image, cv2.COLOR_BGR2BGRA, with_alpha)
+            else:
+                cv2.cvtColor(self.image, cv2.COLOR_GRAY2BGRA, with_alpha)
+            self.image = with_alpha
diff --git a/setup.py b/setup.py
index 6ee84d6..7bba119 100644
--- a/setup.py
+++ b/setup.py
@@ -44,7 +44,9 @@ OpenCV imaging engine for thumbor.
     include_package_data=True,
     install_requires=[
         'colour',
+        'numpy',
         'thumbor',
+        'opencv-python',
     ],
     extras_require={
         'tests': tests_require,
```

**Comment by Arthien** [comment]
:)

**Comment by kkopachev** [comment]
@fmunteanu just curious, if your thumbor serves 10million images per hour, how many images are cached on akamai then? How many ec2 instances you have? 

We have similar stack. Cloudfront in front, then nginx+thumbor instances. S3 for loader and result cache, no local caches. We serve around 500k/hour (reaching thumbor), much more cached on cloudfront.

We did simple math and realized having Fargate would be more expensive than running on reserved instances, but then we switched to Spot instances and now it's much more cheaper. We use Fleet of Spot instances across 3 AZs with 10 different instance types. That works out really well, since thumbor is stateless for us and we can handle instance termination within 2 minutes window.

**Comment by fmunteanu** [comment]
@kkopachev we are also planning to switch to Cloudfront, to save network costs. I apologize, I was not clear. We are doing 10,000,000 images/hour hitting Akamai and 3,000,000 images/hour hitting the Thumbor cluster. The offload is very low, we evaluated originally a 98% offload, but we are doing approx 70%. What is your Cloudfront offload?

**Comment by Savar** [comment]
@fmunteanu while doing this with this setup, how many instances (aka CPUs of which type) are you using?

Also did you ever evaluate to use opencv instead of pillow? We actually have to still start doing autoscaling but we use quite a bunch of machines right now without even doing the detection. So I'm curious how big your cluster becomes with 3 million images/hour..

**Comment by fmunteanu** [comment]
@Savar we switched 2 days ago to OpenCV 3.4.1 and the `opencv-engine` CV3 engine made by @Arthien. The differences are dramatic, we passed from 42 instances at peak time, to 18. We are currently using `c5.2xlarge` CentOS7 instances, they behave at their best with the `RequestCountPerTarget` metric. It acts like a barometer, allowing a very smooth autoscaling. I found that in Expedia's unpredictable bursting traffic, scaling-in/out based on CPU was not efficient.

**Comment by fmunteanu** [comment]
@Savar are you running on Amazon Linux 1? I prefer CentOS7, I ported `aws-cfn-bootstrap-1.4` to it, so we can properly signal and also have full systemd support. I run this shell with Packer, to build a custom AMI:
```
declare -A rpm
rpm[10]='aws-cfn-bootstrap'
rpm[11]='golang-github-prometheus-node_exporter'
# more packages here...
sudo /bin/yum -y install http://rpm.axivo.com/axivo-release-7-1.noarch.rpm
sudo /bin/yum --enablerepo=axivo -y install ${rpm[@]}
unset rpm
```

I find CentOS7.4 very robust.

**Comment by kkopachev** [comment]
@fmunteanu not sure how much CloudFront offloads, since it's shared distribution with lots of prefixes. 
How did you chose `ENGINE_THREADPOOL_SIZE` value to be `2 x nproc`? Is there an explanation behind this? 

Do you guys see any quality loss when using OpenCV engine? According to https://python-pillow.github.io/pillow-perf/ OpenCV does not have fair convolution resampling filter, so comparing it to stock pil engine would be, well, not fair. Or I am missing something here? Maybe there is a little quality loss?
Also it looks weird that in pillow-perf tests show OpenCV has lower speeds at load/save of jpegs (most of images anyway?)  compared to mainstream pillow. Isn't that suspicious for you guys?

**Comment by kkopachev** [comment]
As @Savar pointed out, that might be very setup-dependent on how much CDN can offload. 
I parsed CF logs for one hour and got these numbers:
1782634 Hit / 194500 Miss (per hour)

CloudFront does not share cache between edge locations (IIRC) so each new location would have to go back to origin to fetch resource. Not quite sure if Akamai supports shared cache.

**Comment by fmunteanu** [comment]
@kkopachev Akamai does not support a shared cache structure either. Basically you have the Edge servers hitting the Parent servers (second cache layer) which finally hit the Origin servers (Thumbor).

**Comment by heynemann** [comment]
@fmunteanu I was not advocating you use openCV for your main engine as we use PIL at globo.com. I was advocating you use RemoteCV for your object detection as we do.

**Comment by Savar** [comment]
@fmunteanu do you have an example image (big, resized with pillow-simd and resized with opencv) which you could share? I am curious because I wasn't aware of the scaling issues opencv might have.

**Comment by gingerlime** [comment]
I'm a bit late to the party, and to be fair, I'm dealing with much lower volume currently, but have you considered using a local disk-based caching as well? We're using docker-based [nginx-proxy](https://github.com/MinimalCompact/thumbor/tree/master/nginx-proxy) and disk-based results storage, and it drops a huge load off of thumbor, independently of our CDN (cloudflare in our case). Of course, the more servers you have, and assuming they can't easily share storage, then you'll still get misses, but any cache hit can still save considerable processing time. I assume your local instances are still going to be a smaller number than CDN edges, so it's still more likely to be a win.
